### PR TITLE
[@types/slate] Update for version 0.47

### PR DESCRIPTION
**Pull request by kaydelaney**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/ianstormtaylor/slate/tree/67e397100e2f4cc6bcd4b8f98d7a85dcf30c72fd/packages/slate/src
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

Making this as a draft PR since there are a lot of changes and I'd like to get some more opinions before making a proper PR.

Besides updating the definitions to match version 0.47 of Slate, I've also added some definitions that were missing before, and changed definitions that weren't totally correct.

### CODE REVIEW

**Comment by benjiro** [comment]
@kaydelaney Thanks, I should have some spare time on the weekend to review it.

Cheers,
Ben

**Comment by kaydelaney** [comment]
@benjiro  awesome, thanks!

File: types/slate/index.d.ts
Comment (fuafa) [inline types/slate/index.d.ts]: Should this go to `SchemaProperties`?

Comment (kaydelaney) [reply types/slate/index.d.ts]: Yes it should, nice catch, thanks!

**Review by benjiro — APPROVED** [review]
LGTM 
Thanks for all the work @kaydelaney, the changes look good to me.

**Comment by benjiro** [comment]
Don't merge this yet, please.
@fuafa Once you've finished reviewing and ok with the changes, we will merge this in.

**Review by benjiro — CHANGES_REQUESTED** [review]
#88 RenderMark has now become  RenderAnnotation & RenderDecoration
https://github.com/ianstormtaylor/slate/blob/master/packages/slate-react/Changelog.md

```
The renderMark middleware is no longer used for decorations! Previously renderMark would be used for rendering both types of decorations. Now there are separate renderAnnotation and renderDecoration middleware functions to use instead.
```

**Comment by kaydelaney** [comment]
> #88 RenderMark has now become RenderAnnotation & RenderDecoration
> https://github.com/ianstormtaylor/slate/blob/master/packages/slate-react/Changelog.md
> 
> ```
> The renderMark middleware is no longer used for decorations! Previously renderMark would be used for rendering both types of decorations. Now there are separate renderAnnotation and renderDecoration middleware functions to use instead.
> ```

@benjiro Just pushed a commit (hopefully) fixing this. Also added utility methods to the Editor class.

**Review by benjiro — APPROVED** [review]
Thanks for that change, looks good now. I'm right for this to be merged

**Comment by kaydelaney** [comment]
@benjiro awesome!

**Comment by jessetrinity** [comment]
Before I merge this, do you want to be updating the types version to 0.47 (currently at 0.22) to match the library?

**Comment by kaydelaney** [comment]
The package `slate` is at version 0.47, but `slate-react` is at 0.22 as far as I know, and the versions in the respective `index.d.ts` files reflect that I think unless I'm missing something?

**Comment by jessetrinity** [comment]
Nope, I was looking at the title that says @types/slate." The file indeed says slate-react.

**Comment by DinMon** [comment]
Hey just wanna check the type SlateNode which is derived from Node does not match with function getEventTransfer(event) return type which is an object containting { files, fragment, html, node, rich, text }. Please could you check if the type SlateNode should be the returning data type of getEventTransfer?

**Pull request merged**
